File: python/bvh.py

from math import inf
import numpy as np
import random
from python.hittables import HittableList
from sphere import AxisAlignedBoundingBox, Hittable, Ray, Vector, surrounding_box, HitRecord


class BoundingVolumeHierarchyNode(Hittable):

    def __init__(self, object_list: HittableList, time0: float, time1: float, start: int = -1, end: int = -1):

        if start == -1:
            start = 0
        if end == -1:
            end = len(object_list.hittable_objects)

        axis = random_int(0, 2)

        def sort_key(box):
            return get_box_min(box, axis)

        objects_span = end - start

        if objects_span == 1:
            self.left = object_list.hittable_objects[start]
            self.right = self.left

        elif objects_span == 2:
            if sort_key(object_list.hittable_objects[start]) < sort_key(object_list.hittable_objects[start+1]):
                self.left = object_list.hittable_objects[start]
                self.right = object_list.hittable_objects[start+1]
            else:
                self.left = object_list.hittable_objects[start+1]
                self.right = object_list.hittable_objects[start]

        else:
            sorted_list = sorted(object_list.hittable_objects, key=sort_key)
            object_list = HittableList(sorted_list[0])

            for i in range(1, len(sorted_list)):
                object_list.add(sorted_list[i])

            mid = start + objects_span//2
            self.left = BoundingVolumeHierarchyNode(
                object_list, time0, time1, start, mid)
            self.right = BoundingVolumeHierarchyNode(
                object_list, time0, time1, mid, end)

        box_left = self.left.bounding_box(time0, time1)
        box_right = self.right.bounding_box(time0, time1)
        if box_left == False or box_right == False:
            box_left = self.left.box
            box_right = self.right.box

        self.box = surrounding_box(box_left, box_right)

    def is_hit(self, ray: Ray, t_min: float, t_max: float, t_hit):
        if not self.box.is_hit(ray, t_min, t_max, t_hit):
            return False

        t_hit_left = np.zeros(1)
        hit_left = self.left.is_hit(ray, t_min, t_max, t_hit_left)
        t_hit_right = np.zeros(1)
        hit_right = self.right.is_hit(ray, t_min, t_max, t_hit_right)

        if hit_left and not hit_right:
            t_hit[0] = t_hit_left[0]
        elif hit_right and not hit_left:
            t_hit[0] = t_hit_right[0]
        elif hit_left and hit_right:
            if t_hit_left < t_hit_right:
                t_hit[0] = t_hit_left[0]
            else:
                t_hit[0] = t_hit_right[0]

        return hit_left or hit_right

    def get_hit_record(self, ray: Ray, t_hit: float) -> HitRecord:
        t_hit_left = np.zeros(1)
        hit_left = self.left.is_hit(ray, 0, inf, t_hit_left)
        t_hit_right = np.zeros(1)
        hit_right = self.right.is_hit(ray, 0, inf, t_hit_right)

        if hit_left and not hit_right:
            return self.left.get_hit_record(ray, t_hit_left[0])
        elif hit_right and not hit_left:
            return self.right.get_hit_record(ray, t_hit_right[0])
        elif hit_left and hit_right:
            if t_hit_left < t_hit_right:
                return self.left.get_hit_record(ray, t_hit_left[0])
            else:
                return self.right.get_hit_record(ray, t_hit_right[0])

        return None


def get_box_min(hittable: Hittable, axis: int) -> float:
    if hittable.has_box():
        return hittable.bounding_box(0, 0).minimum[axis]


def random_int(min: int, max: int):
    return int(random.random()*(max-min) - min)

File: cpp/hittables/bvh.cpp

#include "bvh.h"

/**
 * @brief      Constructs the BHV tree.
 * @param[in]  objects  The objects
 * @param[in]  start    The start index
 * @param[in]  end      The end index
 * @param[in]  time0    The time 0
 * @param[in]  time1    The time 1
 */
BVHNode::BVHNode(std::vector<std::shared_ptr<Hittable>> &objects, size_t start, size_t end, double time0, double time1)
{
    int axis = random_int(0, 2);
    auto comparator = (axis == 0)   ? box_x_compare
                      : (axis == 1) ? box_y_compare
                                    : box_z_compare;
    size_t object_span = end - start;

    if (object_span == 1)
    {
        left = right = objects[start];
    }
    else if (object_span == 2)
    {
        if (comparator(objects[start], objects[start + 1]))
        {
            left = objects[start];
            right = objects[start + 1];
        }
        else
        {
            left = objects[start + 1];
            right = objects[start];
        }
    }
    else
    {
        std::sort(objects.begin() + start, objects.begin() + end, comparator);

        auto mid = start + object_span / 2;
        left = std::make_shared<BVHNode>(objects, start, mid, time0, time1);
        right = std::make_shared<BVHNode>(objects, mid, end, time0, time1);
    }

    AABB box_left, box_right;

    if (!left->bounding_box(time0, time1, box_left) || !right->bounding_box(time0, time1, box_right))
    {
        std::cerr << "No bounding box in BVHNode constructor." << std::endl;
    }

    box = surrounding_box(box_left, box_right);
}

/**
 * @brief      Checks if the ray hits the bounding box of the node.
 * @param[in]  r       The ray
 * @param[in]  t_min   The minimum parameter value
 * @param[in]  t_max   The maximum parameter value
 * @param[out] rec     The hit record
 * @return     True if the ray hits the BVH node bounding box, False otherwise.
 */
bool BVHNode::hit(const Ray &r, double t_min, double t_max, HitRecord &rec) const
{
    if (!box.hit(r, t_min, t_max))
    {
        return false;
    }

    bool hit_left = left->hit(r, t_min, t_max, rec);
    bool hit_right = right->hit(r, t_min, hit_left ? rec.get_t() : t_max, rec);

    return hit_left || hit_right;
}

/**
 * @brief      Get the bounding box of the node.
 * @param[in]  time0       The time 0
 * @param[in]  time1       The time 1
 * @param[out]      output_box  The output box
 * @return     True if the bounding box is valid, False otherwise.
 */
bool BVHNode::bounding_box(double time0, double time1, AABB &output_box) const
{
    output_box = box;
    return true;
}

/**
 * @brief      Compares the bounding boxes of two objects.
 * @param[in]  a      The first object
 * @param[in]  b      The second object
 * @param[in]  axis   The axis
 * @return     True if the bounding box of the first object is smaller than the bounding box of the second object, False otherwise.
 */
bool box_compare(const std::shared_ptr<Hittable> a, const std::shared_ptr<Hittable> b, int axis)
{
    AABB box_a;
    AABB box_b;

    if (!a->bounding_box(0, 0, box_a) || !b->bounding_box(0, 0, box_b))
    {
        std::cerr << "No bounding box in bvh_node constructor." << std::endl;
    }

    return box_a.get_min().e[axis] < box_b.get_min().e[axis];
}

/**
 * @brief      Compares the bounding boxes of two objects along the x axis.
 * @param[in]  a      The first object
 * @param[in]  b      The second object
 * @return     True if the bounding box of the first object is smaller than the bounding box of the second object, False otherwise.
 */

bool box_x_compare(const std::shared_ptr<Hittable> a, const std::shared_ptr<Hittable> b)
{
    return box_compare(a, b, 0);
}

/**
 * @brief      Compares the bounding boxes of two objects along the y axis.
 * @param[in]  a      The first object
 * @param[in]  b      The second object
 * @return     True if the bounding box of the first object is smaller than the bounding box of the second object, False otherwise.
 */
bool box_y_compare(const std::shared_ptr<Hittable> a, const std::shared_ptr<Hittable> b)
{
    return box_compare(a, b, 1);
}

/**
 * @brief      Compares the bounding boxes of two objects along the z axis.
 * @param[in]  a      The first object
 * @param[in]  b      The second object
 * @return     True if the bounding box of the first object is smaller than the bounding box of the second object, False otherwise.
 */
bool box_z_compare(const std::shared_ptr<Hittable> a, const std::shared_ptr<Hittable> b)
{
    return box_compare(a, b, 2);
}
